Jabref corrupts PDF file when writing XMP metadata to linked PDF

### JabRef version

Latest development branch build (please note build date below)

### Operating system

Windows

### Details on version and operating system

JabRef 5.6--2022-04-04--dbf921e Windows 10 10.0 amd64  Java 17.0.2  JavaFX 18+12

### Checked with the latest development build

- [X] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

1. Have a linked pdf file
   (my pdf file was 838 kb)
2. Have bibliographic data
3. Write XMP metadata to pdf via `Tools > write XMP metadata to PDF files (F6)`

What happens:

- Trying to open this pdf file with a pdf reader (e.g. foxit reader / pdf xchange) shows the pdf has no pages anymore.
- The size of the pdf shrinks to ~ 4kb

### Appendix

_No response_

I also encountered an error message just now when doing some other test trying to write XMP metadata to linked pdf.

But that was related to groups and this entrytable/groups index bug. Maybe not related to this issue here. I removed a keyword, hence the entry was removed from a group. Then I tried to write the XMP metadata.

```
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalStateException: 1-based index not found: 1
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache$Segment.get(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache.get(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache.getOrLoad(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache$LocalLoadingCache.get(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(Unknown Source)
	at org.jabref@5.6.144/org.jabref.logic.pdf.FileAnnotationCache.getFromCache(Unknown Source)
	at org.jabref@5.6.144/org.jabref.gui.entryeditor.fileannotationtab.FileAnnotationTabViewModel.lambda$reloadAnnotations$1(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.win.WinApplication._enterNestedEventLoopImpl(Native Method)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.win.WinApplication._enterNestedEventLoop(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.Application.enterNestedEventLoop(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.EventLoop.enter(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.tk.quantum.QuantumToolkit.enterNestedEventLoop(Unknown Source)
	at org.jabref.merged.module@5.6.144/javafx.stage.Stage.showAndWait(Unknown Source)
	at org.jabref.merged.module@5.6.144/javafx.scene.control.HeavyweightDialog.showAndWait(Unknown Source)
	at org.jabref.merged.module@5.6.144/javafx.scene.control.Dialog.showAndWait(Unknown Source)
	at org.jabref@5.6.144/org.jabref.gui.JabRefDialogService.showErrorDialogAndWait(Unknown Source)
	at org.jabref@5.6.144/org.jabref.gui.FallbackExceptionHandler.lambda$uncaughtException$0(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.win.WinApplication._enterNestedEventLoopImpl(Native Method)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.win.WinApplication._enterNestedEventLoop(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.Application.enterNestedEventLoop(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.EventLoop.enter(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.tk.quantum.QuantumToolkit.enterNestedEventLoop(Unknown Source)
	at org.jabref.merged.module@5.6.144/javafx.stage.Stage.showAndWait(Unknown Source)
	at org.jabref.merged.module@5.6.144/javafx.scene.control.HeavyweightDialog.showAndWait(Unknown Source)
	at org.jabref.merged.module@5.6.144/javafx.scene.control.Dialog.showAndWait(Unknown Source)
	at org.jabref@5.6.144/org.jabref.gui.JabRefDialogService.showErrorDialogAndWait(Unknown Source)
	at org.jabref@5.6.144/org.jabref.gui.FallbackExceptionHandler.lambda$uncaughtException$0(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at org.jabref.merged.module@5.6.144/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalStateException: 1-based index not found: 1
	at org.jabref.merged.module@5.6.144/org.apache.pdfbox.pdmodel.PDPageTree.get(Unknown Source)
	at org.jabref.merged.module@5.6.144/org.apache.pdfbox.pdmodel.PDPageTree.get(Unknown Source)
	at org.jabref@5.6.144/org.jabref.logic.pdf.PdfAnnotationImporter.importAnnotations(Unknown Source)
	at org.jabref@5.6.144/org.jabref.logic.pdf.EntryAnnotationImporter.lambda$importAnnotationsFromFiles$2(Unknown Source)
	at java.base/java.util.Optional.ifPresent(Unknown Source)
	at org.jabref@5.6.144/org.jabref.logic.pdf.EntryAnnotationImporter.importAnnotationsFromFiles(Unknown Source)
	at org.jabref@5.6.144/org.jabref.logic.pdf.FileAnnotationCache$1.load(Unknown Source)
	at org.jabref@5.6.144/org.jabref.logic.pdf.FileAnnotationCache$1.load(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache$Segment.loadSync(Unknown Source)
	at org.jabref.merged.module@5.6.144/com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(Unknown Source)
	... 42 more
```

--- a/src/main/java/org/jabref/logic/exporter/EmbeddedBibFilePdfExporter.java
+++ b/src/main/java/org/jabref/logic/exporter/EmbeddedBibFilePdfExporter.java
@@ -119,7 +119,7 @@ public class EmbeddedBibFilePdfExporter extends Exporter {
                 nameDictionary.setEmbeddedFiles(efTree);
                 document.getDocumentCatalog().setNames(nameDictionary);
             }
-            Path newFile = Path.of(file.toAbsolutePath().toString() + ".copy");
+            Path newFile = Files.createTempFile("JabRef", "pdf");
             document.save(newFile.toFile());
             FileUtil.copyFile(newFile, file, true);
             Files.delete(newFile);

--- a/src/main/java/org/jabref/logic/exporter/EmbeddedBibFilePdfExporter.java
+++ b/src/main/java/org/jabref/logic/exporter/EmbeddedBibFilePdfExporter.java
@@ -68,6 +68,7 @@ public class EmbeddedBibFilePdfExporter extends Exporter {
         if (!Files.exists(file) || !FileUtil.isPDFFile(file)) {
             return;
         }
+
         try (PDDocument document = Loader.loadPDF(file.toFile())) {
             PDDocumentNameDictionary nameDictionary = document.getDocumentCatalog().getNames();
             PDEmbeddedFilesNameTreeNode efTree;
@@ -118,7 +119,10 @@ public class EmbeddedBibFilePdfExporter extends Exporter {
                 nameDictionary.setEmbeddedFiles(efTree);
                 document.getDocumentCatalog().setNames(nameDictionary);
             }
-            document.save(file.toFile());
+            Path newFile = Path.of(file.toAbsolutePath().toString()+ ".copy");
+            document.save(newFile.toFile());
+            FileUtil.copyFile(newFile, file, true);
+            Files.delete(newFile);
         }
     }
 

--- a/src/main/java/org/jabref/logic/exporter/EmbeddedBibFilePdfExporter.java
+++ b/src/main/java/org/jabref/logic/exporter/EmbeddedBibFilePdfExporter.java
@@ -119,7 +119,7 @@ public class EmbeddedBibFilePdfExporter extends Exporter {
                 nameDictionary.setEmbeddedFiles(efTree);
                 document.getDocumentCatalog().setNames(nameDictionary);
             }
-            Path newFile = Path.of(file.toAbsolutePath().toString()+ ".copy");
+            Path newFile = Path.of(file.toAbsolutePath().toString() + ".copy");
             document.save(newFile.toFile());
             FileUtil.copyFile(newFile, file, true);
             Files.delete(newFile);

--- a/src/main/java/org/jabref/logic/xmp/XmpUtilReader.java
+++ b/src/main/java/org/jabref/logic/xmp/XmpUtilReader.java
@@ -112,7 +112,7 @@ public class XmpUtilReader {
      *
      * @return empty List if no metadata has been found, or cannot properly find start or end tag in metadata
      */
-    private static List<XMPMetadata> getXmpMetadata(PDDocument document) throws IOException {
+    private static List<XMPMetadata> getXmpMetadata(PDDocument document) {
         PDDocumentCatalog catalog = document.getDocumentCatalog();
         PDMetadata metaRaw = catalog.getMetadata();
         List<XMPMetadata> metaList = new ArrayList<>();

--- a/src/main/java/org/jabref/logic/xmp/XmpUtilWriter.java
+++ b/src/main/java/org/jabref/logic/xmp/XmpUtilWriter.java
@@ -5,6 +5,7 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.UnsupportedEncodingException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -15,6 +16,7 @@ import java.util.stream.Collectors;
 
 import javax.xml.transform.TransformerException;
 
+import org.jabref.logic.util.io.FileUtil;
 import org.jabref.model.database.BibDatabase;
 import org.jabref.model.entry.BibEntry;
 import org.jabref.model.entry.field.Field;
@@ -328,7 +330,10 @@ public class XmpUtilWriter {
 
             // Save
             try {
-                document.save(path.toFile());
+                Path newFile = Path.of(path.toAbsolutePath().toString() + ".copy");
+                document.save(newFile.toFile());
+                FileUtil.copyFile(newFile, path, true);
+                Files.delete(newFile);
             } catch (IOException e) {
                 LOGGER.debug("Could not write XMP metadata", e);
                 throw new TransformerException("Could not write XMP metadata: " + e.getLocalizedMessage(), e);

--- a/src/main/java/org/jabref/logic/xmp/XmpUtilWriter.java
+++ b/src/main/java/org/jabref/logic/xmp/XmpUtilWriter.java
@@ -330,7 +330,7 @@ public class XmpUtilWriter {
 
             // Save
             try {
-                Path newFile = Path.of(path.toAbsolutePath().toString() + ".copy");
+                Path newFile = Files.createTempFile("JabRef", "pdf");
                 document.save(newFile.toFile());
                 FileUtil.copyFile(newFile, path, true);
                 Files.delete(newFile);

